hypex/comparators/abstract.py

```diff
@@ -125,41 +138,164 @@ def _extract_dataset(
+            ValueError: If `compare_by` is not one of the allowed values.
+        """
+        if compare_by == "groups":
+            if isinstance(target_fields, List) and len (target_fields) > 1:
+                raise ValueError("Only one Target field can be passed when the comparison is done by groups")
+        elif baseline_field is None:
+            raise NoRequiredArgumentError("baseline_field")
+
+        if compare_by != "columns":
+            if group_field is None:
+                raise NoRequiredArgumentError("group_field")
+            elif isinstance(group_field, List):
+                group_field = Adapter.list_to_single(group_field)
+            if not isinstance(group_field, str):
+                raise TypeError(
+                    f"group_field must be one string, {type(group_field)} passed."
+                )
+
+        target_fields = Adapter.to_list(target_fields)
+        if compare_by == "groups":
+            data_buckets = data.groupby(by=group_field, fields_list=target_fields)
+            baseline_data = cls._split_ds_into_columns([data_buckets.pop(0)])
+            compared_data = cls._split_ds_into_columns(data=data_buckets)
+        elif compare_by == "columns":
+            baseline_data = [
+                (f"0{NAME_BORDER_SYMBOL}{baseline_field}", data[baseline_field])
+            ]
+            compared_data = [
+                (f"0{NAME_BORDER_SYMBOL}{column}", data[column])
+                for column in target_fields
+            ]
+        elif compare_by == "columns_in_groups":
+            baseline_data = cls._split_ds_into_columns(
+                data.groupby(by=group_field, fields_list=baseline_field)
+            )
+            compared_data = data.groupby(by=group_field, fields_list=target_fields)
+            compared_data = cls._split_ds_into_columns(data=compared_data)
+        elif compare_by == "cross":
+            data_buckets = data.groupby(by=group_field, fields_list=baseline_field)
+            baseline_data = cls._split_ds_into_columns([data_buckets.pop(0)])
+            compared_data = data.groupby(by=group_field, fields_list=target_fields)
+            compared_data.pop(0)
+            compared_data = cls._split_ds_into_columns(data=compared_data)
+        else:
+            raise ValueError("compare_by")
+        return baseline_data, compared_data
+
+    @classmethod
+    def calc(
+        cls,
+        data: Dataset,
+        compare_by: Literal["groups", "columns", "columns_in_groups", "cross"],
+        target_fields: Union[str, List[str], None] = None,
+        baseline_field: Optional[str] = None,
+        group_field: Optional[str] = None,
+        grouping_data: Optional[Tuple[List[Tuple[str, Dataset]]]] = None,
+        **kwargs,
+    ) -> Dict:
+        target_fields = Adapter.to_list(target_fields)
+        baseline_field = Adapter.list_to_single(baseline_field)
+        group_field = Adapter.list_to_single(group_field)
```

Comment:
what is the reason of this action if u have typing Optional[str]?

Reply:
Because Optional[str] does not prevent users from passing a list in fact